### PR TITLE
New implementation for replace

### LINKED ISSUE #570: performance issue with replace

### PR DESCRIPTION
**Pull request by GerardPaligot**
Closes #570

### CODE REVIEW

**Comment by monperrus** [comment]
looks good. why did you put there the commits related to the output?

**Comment by monperrus** [comment]
I understand. Because you put the result in src/main/java. It really makes sense.

Could you separate those commits in a PR that we would accept first?

**Comment by GerardPaligot** [comment]
Which commits?

**Comment by monperrus** [comment]
https://github.com/INRIA/spoon/pull/632/commits/79803cba3bf75baf8f1520b76d42a059a80732aa

https://github.com/INRIA/spoon/pull/632/commits/4fb7228283300e08b185cceb3120523bbfd1d669

Others are related to the formatting changes?

--Martin

**Comment by GerardPaligot** [comment]
PR done here #634

**Comment by monperrus** [comment]
merged, this one can be rebased.

File: src/main/java/spoon/generating/replace/ReplacementVisitor.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+import spoon.SpoonException;
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.visitor.CtScanner;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+class ReplacementVisitor extends CtScanner {
```
Comment (GerardPaligot) [inline src/main/java/spoon/generating/replace/ReplacementVisitor.java]: Adds javadoc.

File: src/test/java/spoon/test/api/APITest.java [hunk 2]
```diff
@@ -260,4 +278,85 @@ public void testInvalidateCacheOfCompiler() throws Exception {
 
 		assertTrue(spoon.getModelBuilder().compile());
 	}
+
+	@Test
+	public void testSetterInNodes() throws Exception {
```
Comment (GerardPaligot) [inline src/test/java/spoon/test/api/APITest.java]: Adds contract.

Comment (GerardPaligot) [inline src/test/java/spoon/test/api/APITest.java]: Done.

File: src/main/java/spoon/generating/GeneratingTypeProcessor.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import spoon.reflect.reference.CtTypeReference;
+import spoon.reflect.visitor.CtScanner;
+import spoon.reflect.visitor.ReferenceFilter;
+
+import java.util.List;
+
+import static spoon.generating.replace.ReplaceScanner.GENERATING_REPLACE_VISITOR;
+import static spoon.generating.replace.ReplaceScanner.TARGET_REPLACE_PACKAGE;
+
+public class GeneratingTypeProcessor extends AbstractProcessor<CtType<?>> {
```
Comment (GerardPaligot) [inline src/main/java/spoon/generating/GeneratingTypeProcessor.java]: Change to ReplacementVisitorGenerator

File: src/main/java/spoon/support/reflect/declaration/CtElementImpl.java [hunk 4]
```diff
@@ -228,7 +224,7 @@ public boolean equals(Object o) {
 	}
 
 	public List<CtAnnotation<? extends Annotation>> getAnnotations() {
-		return Collections.unmodifiableList(annotations);
+		return annotations;
```
Comment (monperrus) [inline src/main/java/spoon/support/reflect/declaration/CtElementImpl.java]: why this?

File: src/main/java/spoon/support/reflect/reference/CtExecutableReferenceImpl.java [hunk 5]
```diff
@@ -181,7 +180,7 @@ public boolean isConstructor() {
 
 	@Override
 	public List<CtTypeReference<?>> getParameters() {
-		return Collections.unmodifiableList(parameters);
+		return parameters;
```
Comment (monperrus) [inline src/main/java/spoon/support/reflect/reference/CtExecutableReferenceImpl.java]: idem

File: src/main/java/spoon/support/reflect/declaration/CtElementImpl.java (hunk 4 above)
Comment (GerardPaligot) [inline src/main/java/spoon/support/reflect/declaration/CtElementImpl.java]: See

https://github.com/INRIA/spoon/pull/632/files#diff-687e64c373505dec021ad8918e3086e1R49
https://github.com/INRIA/spoon/pull/632/files#diff-687e64c373505dec021ad8918e3086e1R68
https://github.com/INRIA/spoon/pull/632/files#diff-687e64c373505dec021ad8918e3086e1R85

**Comment by GerardPaligot** [comment]
@monperrus Should be ok here.

File: src/test/java/spoon/reflect/visitor/CtScannerTest.java [hunk 6]
```diff
@@ -42,4 +53,47 @@ public void testScannerContract() throws Exception {
...
+		launcher.addInputResource("./src/main/java/spoon/reflect/internal");
+		// Utils.
+		launcher.addInputResource("./src/main/java/spoon/reflect/visitor/CtScanner.java");
+		launcher.addInputResource("./src/main/java/spoon/generating/replace/");
+		launcher.addProcessor(new ReplacementVisitorGenerator());
+		launcher.setOutputFilter(new RegexFilter("spoon.support.visitor.replace.*"));
+		launcher.run();
+
+		assertThat(build(new File("./src/main/java/spoon/support/visitor/replace/ReplacementVisitor.java")).Class().get(ReplacementVisitor.class))
+				.isEqualTo(build(new File("./target/generated/spoon/support/visitor/replace/ReplacementVisitor.java")).Class().get(ReplacementVisitor.class).toString());
```
Comment (monperrus) [inline src/test/java/spoon/reflect/visitor/CtScannerTest.java]: .toString() required?

**Comment by GerardPaligot** [comment]
@monperrus Ok here.

**Pull request merged**